### PR TITLE
Add build vit models notes

### PR DESCRIPTION
**Pull request by rentainhe**
## TODO
- [x] 添加关于如何implement ViT模型的Notes

### CODE REVIEW

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 1]
```diff
@@ -0,0 +1,434 @@
+# How to build vision transformer model in LiBai
+It's easy for the users to use libai's built-in layers, e.g., [libai.layers](https://libai.readthedocs.io/en/latest/modules/libai.layers.html), to build the `transformer-based` models. Let's take a deep look into the process of building vision transformer model in LiBai.
```
Comment (Ldpe2G) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:2]:
```suggestion
It's easy for the users to build the `transformer-based` models by using the libai's built-in [layers](https://libai.readthedocs.io/en/latest/modules/libai.layers.html). Let's take a deep look into the process of building vision transformer model in LiBai.
```

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+## Model Architecture
+**Vision Transformer** was released with paper [An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale](https://arxiv.org/abs/2010.11929) by Alexey Dosovitskiy, Lucas Beyer, Alexander Kolesnikov, Dirk Weissenborn, Xiaohua Zhai, Thomas Unterthiner, Mostafa Dehghani, Matthias Minderer, Georg Heigold, Sylvain Gelly, Jakob Uszkoreit, Neil Houlsby.
+
+**Vision Transformer** model contains three parts: `Patch Embedding` + `Transformer Block` + `Linear Classification Head`, which can be summarized as the following picture:
+
+![](./assets/vision_transformer.png)
+
+## A simple torch implementation of Vision Transformer
+Let's take a look at the pytorch implementation of ViT models, this code is modified from [timm.models.vision_transformer](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py):
```
Comment (Ldpe2G) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:12]:
```suggestion
The following code shows the pytorch implementation of ViT models which is modified from [timm.models.vision_transformer](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py):
```

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+        x = self.norm(x)
+        outcome = x[:, 0]
+        return self.head(outcome)
+    
+    def forward(self, x):
+        x = self.forward_features(x)
+        x = self.forward_head(x)
+        return x
+```
+We further decoupled the forward function into `forward_features` and `forward_head`:
```
Comment (Ldpe2G) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:185]:
```suggestion
We have further decoupled the forward function into `forward_features` and `forward_head`:
```

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+- `forward_features`: extract the image features using `patch_embed` layer and a stack of `transformer` block
+- `forward_head`: take the `cls_token` of each sample and use `nn.Linear` for classification
+
+## Implement 3D parallel Vision Transformer in LiBai
+In this section, we will show users how to use [libai.layers](https://libai.readthedocs.io/en/latest/modules/libai.layers.html) to build 3D parallel Vision Transformer model with only 100+ lines of code, which is modified from [libai.models.vision_transformer](https://github.com/Oneflow-Inc/libai/blob/main/libai/models/vision_transformer.py)
+
+Here is the LiBai implementation of Vision Transformer models, users only need to replace the pytorch modules with libai.layers as follows:
```
Comment (Ldpe2G) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:192]:
```suggestion
Here is the LiBai implementation of Vision Transformer models, users only need to replace the pytorch modules with the corresponding libai.layers as follows:
```

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+
+LiBai has already implemented `PatchEmbedding`, `TransformerLayer`, `Linear`, `LayerNorm` layers, users only need to replace the module in torch vision transformer models to convert a torch model into libai's style:
+  - `Block` -> `libai.layers.TransformerLayer`
+  - `nn.Linear` -> `libai.layers.Linear`
+  - `nn.LayerNorm` -> `libai.layers.LayerNorm`
+  - `PatchEmbed` -> `libai.layers.PatchEmbedding`
+
+**2. Manually set the sbp signature of `cls_token` and `pos_embed`**
+
+In order to fit different parallel mode in LiBai, users need to manually set the **sbp signature** of the layers built with `nn.Parameter` which are not implemented in LiBai, like `cls_token` and `pos_embed` in Vision Transformer:
```
Comment (Ldpe2G) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:359]:
```suggestion
In order to fit different parallel mode in LiBai, users must manually set the **sbp signature** for all the parameters and buffers of those layers not implemented in LiBai, like `cls_token` and `pos_embed` in Vision Transformer:
```

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md (hunk 1 above)
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:2]: 已改~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md (hunk 2 above)
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:12]: 已改~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md (hunk 3 above)
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:185]: 已改~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md (hunk 4 above)
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:192]: 已改~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md (hunk 5 above)
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:359]: 已改~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+    )
+)
+self.pos_embed = nn.Parameter(
+    flow.zeros(
+        1, num_patches+1, embed_dim,
+        sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
+        placement=dist.get_layer_placement(0),)
+)
+```
+- use `dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast])` to make the `cls_token` and `pos_embed` broadcast in each GPU groups.
```
[the remark this answers is not in the document]
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:379]: 已改~

**Comment by rentainhe** [comment]
## 修复了一下Markdown格式不显示的问题
image
这里原来是个引用，但是不知道为什么readthedocs不支持= =

替换为以下形式:
image

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+
+LiBai has already implemented `PatchEmbedding`, `TransformerLayer`, `Linear`, `LayerNorm` layers, users only need to replace the module in torch vision transformer models to convert a torch model into libai's style:
+  - `Block` -> `libai.layers.TransformerLayer`
+  - `nn.Linear` -> `libai.layers.Linear`
+  - `nn.LayerNorm` -> `libai.layers.LayerNorm`
+  - `PatchEmbed` -> `libai.layers.PatchEmbedding`
+
+**2. Manually set the sbp signature of `cls_token` and `pos_embed`**
+
+In order to fit different parallel mode in LiBai, users must manually set the **sbp signature** for all the parameters and buffers of those layers not implemented in LiBai, like `cls_token` and `pos_embed` in Vision Transformer:
```
Comment (CPFLAME) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:359]: 这里要不要给个SBP的说明文档链接

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+    cls_token = cls_token.to_global(sbp=x.sbp, placement=cls_token.placement)
+    x = flow.cat((cls_token, x), dim=1)
+
+    # use to_global to update the sbp signature of pos_embed
+    pos_embed = self.pos_embed.expand(x.shape[0], -1, -1)
+    pos_embed = pos_embed.to_global(sbp=x.sbp, placement=pos_embed.placement)
+```
+
+**4. Manually set the stage id for pipeline parallel training**
+Most of the built-in layers in LiBai has the arg named `layer_idx` for pipeline parallel settings, to configure a 1F1B pipeline parallel model, users should manually set the stage id for each layers in the models, with the help of `layer_idx` we can simply get a pipeline parallel Vision Transformer models like:
```
Comment (CPFLAME) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:399]: 这里感觉可以再加一句解释一下,  大意应该是  stage id 是为了分配当前的layer到 对应的卡上, 最终形成流水.

Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md]: 已经修改，可以帮忙看看表述是否有问题~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md (hunk 7 above)
Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md]: 已经修改~

File: docs/source/index.rst [hunk 9]
```diff
@@ -10,6 +10,14 @@ Welcome to libai's documentation!
    :maxdepth: 3
 
    tutorials/index
+
+.. toctree::
+   :maxdepth: 2
+
    notes/index
```
Comment (rentainhe) [inline docs/source/index.rst:17]: 这个部分是为了目录里的分级设置做的修改，目前还存在一些line space的问题，这边需要 @lixiang007666 帮忙看看~

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+    # use to_global to update the sbp signature of cls_token
+    cls_token = cls_token.to_global(sbp=x.sbp, placement=cls_token.placement)
+    x = flow.cat((cls_token, x), dim=1)
+
+    # use to_global to update the sbp signature of pos_embed
+    pos_embed = self.pos_embed.expand(x.shape[0], -1, -1)
+    pos_embed = pos_embed.to_global(sbp=x.sbp, placement=pos_embed.placement)
+```
+
+**4. Manually set the stage id for pipeline parallel training**
```
Comment (rentainhe) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:398]: 这个部分的表述再帮忙看看 @L1aoXingyu @CPFLAME @Ldpe2G

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 11]
```diff
@@ -0,0 +1,434 @@
+# Detailed instruction on how to build vision transformer model from scratch in LiBai
```
Comment (rentainhe) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:1]: 在 @khloe-zhang 的建议下修改了一下标题~，和内容更贴切

Comment (L1aoXingyu) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:1]: 这个标题会不会有点太长了

Comment (rentainhe) [reply docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:1]: 好像确实，我改成`Detailed instruction on vision transformer model in LiBai`吧

File: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md [hunk 12]
```diff
@@ -0,0 +1,434 @@
+# Detailed instruction on building vision transformer model in LiBai
```
Comment (rentainhe) [inline docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md:1]: 最终修改的标题~

**Review by Ldpe2G — APPROVED** [review]
